pkg/monitoring/metrics/virt-controller/vmistats_collector.go

```diff
@@ -330,3 +350,69 @@ func collectVMIInterfaceInfo(vmi *k6tv1.VirtualMachineInstance, iface k6tv1.Virt
 		Value: 1.0,
 	}
 }
+
+func collectVMIMigrationTime(vmi *k6tv1.VirtualMachineInstance) []operatormetrics.CollectorResult {
+	var cr []operatormetrics.CollectorResult
+	var migrationName string
+
+	if vmi.Status.MigrationState == nil {
+		return cr
+	}
+
+	migration, exists, err := getMigrationFromMigrationUID(vmi.Namespace, vmi.Status.MigrationState.MigrationUID)
+	if err != nil || !exists {
+		migrationName = none
+	} else {
+		migrationName = migration.Name
+	}
+
+	if vmi.Status.MigrationState.StartTimestamp != nil {
+		cr = append(cr, operatormetrics.CollectorResult{
+			Metric: vmiMigrationStartTime,
+			Value:  float64(vmi.Status.MigrationState.StartTimestamp.Time.Unix()),
+			Labels: []string{vmi.Status.NodeName, vmi.Namespace, vmi.Name, migrationName},
+		})
+	}
+
+	if vmi.Status.MigrationState.EndTimestamp != nil {
+		cr = append(cr, operatormetrics.CollectorResult{
+			Metric: vmiMigrationEndTime,
+			Value:  float64(vmi.Status.MigrationState.EndTimestamp.Time.Unix()),
+			Labels: []string{vmi.Status.NodeName, vmi.Namespace, vmi.Name, migrationName,
+				calculateMigrationStatus(vmi.Status.MigrationState),
+			},
+		})
+	}
+
+	return cr
+}
+
+func calculateMigrationStatus(migrationState *k6tv1.VirtualMachineInstanceMigrationState) string {
+	if !migrationState.Completed {
+		return ""
+	}
+
+	if migrationState.Failed {
+		return "failed"
+	}
+
+	return "succeeded"
+}
+
+func getMigrationFromMigrationUID(namespace string, migrationUID types.UID) (*k6tv1.VirtualMachineInstanceMigration, bool, error) {
+	objs, err := vmiMigrationInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, false, err
+	}
+
+	for _, obj := range objs {
+		curMigration := obj.(*k6tv1.VirtualMachineInstanceMigration)
+		if curMigration.UID != migrationUID {
+			continue
+		}
+
+		return curMigration, true, nil
+	}
+
+	return nil, false, nil
+}
```

Comment:
not sure we need the 2nd (bool) return value. it always false if the first return value is not nil, and vice versa. Can we drop it and check if the first value is not nil, instead?

Reply:
updated